### PR TITLE
[enhancement, CI] add numpydoc validation to pre-commit

### PR DESCRIPTION
**Pull request by icfaust**
## Description

sklearn validates functions and estimators using numpydoc. add this possibly to our linter to get proper styling, not just for patched estimators or functions

just doc changes performance benchmarks unnecessary.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: sklearnex/linear_model/incremental_ridge.py [hunk 1]
```diff
@@ -283,7 +286,7 @@ def fit(self, X, y):
             `n_features` is the number of features. It is necessary for
             `n_samples` to be not less than `n_features` if `fit_intercept`
             is False and not less than `n_features` + 1 if `fit_intercept`
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/incremental_ridge.py:288]:
```suggestion
            ``n_features`` is the number of features. It is necessary for
            ``n_samples`` to be not less than ``n_features`` if ``fit_intercept``
            is False and not less than ``n_features`` + 1 if ``fit_intercept``
```

File: sklearnex/_utils.py [hunk 2]
```diff
@@ -120,11 +120,33 @@ def get_sklearnex_version(rule):
...
+
+    Returns
+    -------
+    decorator : function
+        Function which adds `get_hyperparameters` method to classes.
     """
 
     def decorator(cls):
-        """Add `get_hyperparameters()` static method"""
+        """Add `get_hyperparameters()` static method.
```
Comment (david-cortes-intel) [inline sklearnex/_utils.py:138]:
```suggestion
        """Add ``get_hyperparameters()` static method to a class.

```

File: onedal/utils/validation.py [hunk 3]
```diff
@@ -131,10 +131,18 @@ def _validate_targets(y, class_weight, dtype):
...
 
 def get_finite_keyword():
-    """Gets the argument name for scikit-learn's validation functions compatible with
+    """Return scikit-learn-matching finite check enabling keyword.
+
+    Gets the argument name for scikit-learn's validation functions compatible with
     the current version of scikit-learn and using function inspection instead of
     version check due to `onedal` design rule: sklearn versioning should occur
-    in `sklearnex` module."""
+    in `sklearnex` module.
```
Comment (david-cortes-intel) [inline onedal/utils/validation.py:139]:
```suggestion
    in ``sklearnex`` module.
```

File: onedal/utils/_sycl_queue_manager.py [hunk 4]
```diff
@@ -131,24 +161,32 @@ def from_data(*data):
...
+    ------
+    SyclQueue : SyclQueue or None
+        The global queue after verification.
+
+    Notes
+    -----
+        For most applications, the original queue should be `None`, but
+        if there are nested calls to `manage_global_queue()`, it is
+        important to restore the outer queue, rather than setting it to
+        `None`.
```
Comment (david-cortes-intel) [inline onedal/utils/_sycl_queue_manager.py:189]:
```suggestion
        For most applications, the original queue should be ``None``, but
        if there are nested calls to ``manage_global_queue()``, it is
        important to restore the outer queue, rather than setting it to
        ``None``.
```

File: onedal/primitives/kernel_functions.py [hunk 5]
```diff
@@ -132,21 +167,33 @@ def poly_kernel(X, Y=None, gamma=1.0, coef0=0.0, degree=3, queue=None):
 
 @supports_queue
 def sigmoid_kernel(X, Y=None, gamma=1.0, coef0=0.0, queue=None):
-    """
-    Compute the sigmoid kernel between X and Y:
-        K(x, y) = tanh(scale*dot(x, y^T) + shift)
+    """Compute the sigmoid kernel between X and Y.
+
+    K(x, y) = tanh(gamma*dot(x, y^T) + coef0)
```
Comment (david-cortes-intel) [inline onedal/primitives/kernel_functions.py:172]:
```suggestion
        :math:`K(x, y) = \\text{tanh}(\\gamma \\mathbf{x}^T \\mathbf{y}) + \\text{coef}_0)`
```

File: onedal/linear_model/incremental_linear_model.py [hunk 6]
```diff
@@ -280,14 +283,13 @@ def partial_fit(self, X, y, queue=None):
         return self
 
     def finalize_fit(self, queue=None):
-        """
-        Finalizes ridge regression computation and obtains coefficients
-        from the current `_partial_result`.
+        """Finalize ridge regression from the current `_partial_result`.
```
Comment (david-cortes-intel) [inline onedal/linear_model/incremental_linear_model.py:286]:
```suggestion
        """Finalize ridge regression from the current ``_partial_result``.
```

**Comment by david-cortes-intel** [comment]
@icfaust Just a reminder: the numpydoc format is based off .rst/Sphinx, not markdown.

So wrapping things in a single backtick formats them as _italic_.

Formatting `as code` requires double backticks.

Please use double backticks where appropriate.

File: onedal/basic_statistics/basic_statistics.py [hunk 7]
```diff
@@ -68,15 +68,81 @@ def _get_onedal_params(self, is_csr, dtype=np.float32):
...
 
     @supports_queue
     def fit(self, data, sample_weight=None, queue=None):
+        """Generate statistics.
+
+        Parameters
+        ----------
+        data : array-like of shape (n_samples, n_features)
+            Training data batch, where `n_samples` is the number of samples
+            in the batch, and `n_features` is the number of features.
```
Comment (icfaust) [inline onedal/basic_statistics/basic_statistics.py:131]: @david-cortes-intel what to do in this circumstance for a code snippet? sklearn (where i reference from) does this https://github.com/scikit-learn/scikit-learn/blob/98ed9dc73/sklearn/decomposition/_incremental_pca.py#L211 which yields this in their docs: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.IncrementalPCA.html#sklearn.decomposition.IncrementalPCA.fit

Comment (david-cortes-intel) [reply onedal/basic_statistics/basic_statistics.py:131]: @icfaust Is it possible to reuse the sklearn docstrings without copying them? Like here for example:
https://github.com/uxlfoundation/scikit-learn-intelex/pull/2420

Comment (icfaust) [reply onedal/basic_statistics/basic_statistics.py:131]: good question, unfortunately for basic statistics there is no equivalent. Now that ive re-read your recommendation about ticks, i guess we need to be careful about when a docstring is rendered with markdown vs sphinx. do you have a rule of thumb we should follow to identify when to use a case in a docstring circumstance? i expect we render sphinx for our docs online, but we want to do markdown anywhere else (or am i mistaken?). sorry if these are dumb questions, let me know

Comment (david-cortes-intel) [reply onedal/basic_statistics/basic_statistics.py:131]: @icfaust The docstrings are _never_ rendered as markdown anywhere. The online docs are built using Sphinx, and I understand sklearn itself uses this format even for internal docstrings that are not part of their online docs. Other possible formats besides Sphinx that IDEs would try out would be something like google doc format, which is not markdown either.

Comment (icfaust) [reply onedal/basic_statistics/basic_statistics.py:131]: Thanks for the clarification. In this case, I have a follow up question: How does scikit-learn render the single ticks as code instead of italics?  Along the same route as your suggestion for re-use, wouldn't re-use of single tick sklearn docstrings also cause us issues?  While for our sklearnex-only estimators we define our own docstrings, if we were ever to reuse a method from sklearn defined in a mixin (say TransformerMixin or ClassifierMixin) would it render those mixin method docstrings weirdly because they use single ticks?

Comment (david-cortes-intel) [reply onedal/basic_statistics/basic_statistics.py:131]: Looks like they use a non-default config: https://github.com/scikit-learn/scikit-learn/blob/7a88bf1cd4c05ee0fc1d93d235ddf315378bc0dd/doc/conf.py#L187

Comment (david-cortes-intel) [reply onedal/basic_statistics/basic_statistics.py:131]: Created a PR to mimic their settings: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2466

**Comment by icfaust** [comment]
@david-cortes-intel would you like me still to do the single tick changes? anything else you'd like me to focus on?

File: onedal/common/_backend.py [hunk 8]
```diff
@@ -157,24 +178,30 @@ def bind_default_backend(
...
+        The name of the module where the target function is located (e.g.
+        `covariance`).
+
     lookup_name : Optional[str], optional
-        The name of the method to look up in the backend module. If not provided,
-        the name of the decorated method is used.
+        The name of the method to look up in the backend module. If not
+        provided, the name of the decorated method is used.
+
     no_policy : bool, optional
```
Comment (david-cortes-intel) [inline onedal/common/_backend.py:195]: These look repeated. They could be put into a common docstring like it was done in other cases.

File: onedal/_config.py [hunk 9]
```diff
@@ -59,13 +59,14 @@ def _get_onedal_threadlocal_config():
...
-    """Retrieve current values for configuration set
-    by :func:`sklearnex.set_config`
+    """Retrieve current configuration set by :func:`sklearnex.set_config`
+
     Parameters
     ----------
     copy : bool, default=True
-        If False, a mutable view of the configuration is returned. Each thread
-        has a separate copy of the configuration.
+        If False, a mutable view of the configuration is returned. Each
```
Comment (david-cortes-intel) [inline onedal/_config.py:67]:
```suggestion
        If 'False', a mutable view of the configuration is returned. Each
```

**Review by david-cortes-intel — APPROVED** [review]

**Comment by david-cortes-intel** [comment]
> @david-cortes-intel would you like me still to do the single tick changes? anything else you'd like me to focus on?

No, not needed anymore after the change in config. Left some minor comments.
